ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/StatementExecutorTest.java

```diff
@@ -412,14 +455,185 @@ public void shouldEnforceReferentialIntegrity() {
+    expect(mockEngine.getMetaStore()).andStubReturn(mockMetaStore);
+    expect(mockMetaStore.getSource(name)).andStubReturn(null);
+    expect(mockEngine.buildMultipleQueries(eq(statement), anyObject(), anyObject()))
+        .andReturn(Collections.singletonList(mockQuery));
+    return mockQuery;
+  }
+
+  @Test
+  public void shouldSkipStartForTerminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS";
+    final String terminate = "TERMINATE";
+    final QueryId queryId = new QueryId("csas-query-id");
+    final String name = "foo";
+    final PersistentQueryMetadata mockQuery = mockReplayCSAS(csas, name, queryId);
+    final Statement mockTerminate = mockTerminate(queryId);
+    expect(mockParser.parseSingleStatement(terminate)).andReturn(mockTerminate);
+    expect(mockEngine.terminateQuery(queryId, false)).andReturn(true);
+    expect(mockEngine.getPersistentQueries()).andReturn(Collections.emptySet());
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        ImmutableList.of(
+            new QueuedCommand(
+                new CommandId(Type.STREAM, name, Action.CREATE),
+                new Command(csas, Collections.emptyMap(), Collections.emptyMap())
+            ),
+            new QueuedCommand(
+                new CommandId(Type.TERMINATE, queryId.getId(), Action.EXECUTE),
+                new Command(terminate, Collections.emptyMap(), Collections.emptyMap())
+            )
+        )
+    );
+
+    // Then:
+    verify(mockParser, mockEngine, mockQuery);
+  }
 
+  @Test
+  public void shouldStartUnterminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS ";
+    final List<String> streamNames = ImmutableList.of("foo", "bar", "baz");
+    final List<PersistentQueryMetadata> queries = streamNames.stream()
+        .map(name -> mockReplayCSAS(csas + name, name, new QueryId("query-" + name)))
+        .collect(Collectors.toList());
+    expect(mockEngine.getPersistentQueries()).andStubReturn(queries);
+    for (final PersistentQueryMetadata query : queries) {
+      query.start();
+      expectLastCall();
+    }
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        streamNames.stream()
+            .map(
+                name -> new QueuedCommand(
+                    new CommandId(Type.STREAM, name, Action.CREATE),
+                    new Command(csas + name, Collections.emptyMap(), Collections.emptyMap())
+                ))
+            .collect(Collectors.toList()))
+    ;
+
+    // Then:
+    queries.forEach(EasyMock::verify);
+    verify(mockParser, mockEngine);
+  }
+
+  @Test
+  public void shouldCascade4Dot1DropStreamCommand() {
+    // Given:
+    final String drop = "DROP";
+    final DropStream mockDropStream = mockDropStream("foo");
+    expect(mockParser.parseSingleStatement(drop)).andReturn(mockDropStream);
+    expect(mockEngine.getMetaStore()).andStubReturn(mockMetaStore);
```

Comment:
Do in setup? Just make it optional with `.anyTimes()`. It's just wiring up of mocks, it's not important in the context of this test.